rmw_zenoh_cpp/src/rmw_zenoh.cpp

```diff
@@ -2464,11 +2467,17 @@ rmw_send_request(
   // and any number.
   opts.consolidation = z_query_consolidation_latest();
   opts.value.payload = z_bytes_t{data_length, reinterpret_cast<const uint8_t *>(request_bytes)};
-  client_data->zn_closure_reply =
+  z_owned_closure_reply_t zn_closure_reply =
     z_closure(rmw_zenoh_cpp::client_data_handler, nullptr, client_data);
   z_get(
-    z_loan(context_impl->session), z_loan(
-      client_data->keyexpr), "", &client_data->zn_closure_reply, &opts);
+    z_loan(context_impl->session),
+    z_loan(client_data->keyexpr), "",
+    z_move(zn_closure_reply),
+    &opts);
+
+  // See the comment about the "num_in_flight" class variable in the rmw_client_data_t class for
+  // why we need to do this.
+  client_data->increment_in_flight_callbacks();
```

Comment:
Shouldn't `increment_in_flight_callbacks` be called before `z_get` so that we are sure `rmw_destroy_client` is not called in parallel between `z_get` and `increment_in_flight_callbacks` ? I don't know the rest of the code enough to know if `rmw_destroy_client` and `rmw_send_request`can be called concurrently or not.

Reply:
This is a good point.  We probably actually have a larger problem here in that we don't have any kind of locking at all for concurrent calls of the APIs related to clients (and servers, and subscriptions, and publishers).  That is something we should look into.

For now, I am going to move this `increment` ahead of the `z_get` call as you suggest.